modules/em/src/BayesEM3D.cpp

```diff
@@ -0,0 +1,532 @@
+    }
+  }
+
+  if (IMP::is_nan(x) || IMP::is_nan(y)) {
+    return FloatPair(std::numeric_limits<double>::quiet_NaN(),
+                     std::numeric_limits<double>::quiet_NaN());
+  }
+
+  if (x > y) {
+    return FloatPair(x + log1p(sx * sy * exp(y - x)), sx);
+  } else {
+    return FloatPair(y + log1p(sx * sy * exp(x - y)), sy);
+  }
+}
+
+FloatPair bayesem3d_get_logabssumprodexp(double x, double y, double wx,
+                                         double wy) {
+  double sx = std::copysign(1, wx);
+  double sy = std::copysign(1, wy);
+
+  return bayesem3d_get_logabssumexp(x + log(wx * sx), y + log(wy * sy), sx, sy);
+}
+
+double bayesem3d_get_logsumexp(double x, double y) {
+  if (IMP::isinf(x) || IMP::isinf(y)) {
+    return std::max(x, y);
+  }
+
+  if (IMP::is_nan(x) || IMP::is_nan(y)) {
+    return std::numeric_limits<double>::quiet_NaN();
+  }
+
+  if (x > y) {
+    return x + log1p(exp(y - x));
+  } else {
+    return y + log1p(exp(x - y));
+  }
+}
+
+double bayesem3d_get_logsumprodexp(double x, double y, double wx, double wy) {
+  return bayesem3d_get_logsumexp(x + log(wx), y + log(wy));
+}
+
+std::vector<double> bayesem3d_get_value(Particle *p,
+                                        const algebra::Vector3D &pt,
+                                        double mass_ii,
+                                        const IMP::em::KernelParameters &kps) {
+  std::vector<double> results;
+  algebra::Vector3D cs = core::XYZ(p).get_coordinates();
+
+  double rsq = (cs - pt).get_squared_magnitude();
+  double norm = (kps.get_rnormfac() * mass_ii);
+
+  double tmp = -rsq * kps.get_inv_rsigsq();
+  results.push_back(tmp);
+  results.push_back(norm);
+
+  results.push_back(-2. * (cs[0] - pt[0]) * kps.get_inv_rsigsq());
+  results.push_back(-2. * (cs[1] - pt[1]) * kps.get_inv_rsigsq());
+  results.push_back(-2. * (cs[2] - pt[2]) * kps.get_inv_rsigsq());
+
+  return results;
+}
+
+std::vector<double> bayesem3d_get_value_no_deriv(
+    Particle *p, const algebra::Vector3D &pt, double mass_ii,
+    const IMP::em::KernelParameters &kps) {
+  std::vector<double> results;
+  algebra::Vector3D cs = core::XYZ(p).get_coordinates();
+
+  double rsq = (cs - pt).get_squared_magnitude();
+  double norm = (kps.get_rnormfac() * mass_ii);
+
+  double tmp = -rsq * kps.get_inv_rsigsq();
+  results.push_back(tmp);
+  results.push_back(norm);
+
+  return results;
+}
+
+FloatPair bayesem3d_get_em_density_squared(DensityMap *em,
+                                           long number_of_voxels) {
+  FloatPair results;
+
+  const double *em_data = em->get_data();
+  double em_square = 0;
+  double scale = 0;
+
+  for (long jj = 0; jj < number_of_voxels; ++jj) {
+    em_square += em_data[jj] * em_data[jj];
+    scale += em_data[jj];
+  }
+  results.first = scale;
+  results.second = em_square;
+  return results;
+}
+
+double bayesem3d_get_cross_correlation_coefficient(DensityMap *em1,
+                                                   DensityMap *em2) {
```

Comment:
You might consider marking both `DensityMap*` arguments `const` since it doesn't *look* like you modify either map.